File: android/vbt-app/app/src/main/java/com/vbt/app/data/ble/HeartRateManager.kt
package com.vbt.app.data.ble

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Współdzielony (singleton Hilt) manager pasa/opaski tętna korzystający ze
 * standardowego BLE Heart Rate Service. ConnectViewModel łączy/rozłącza,
 * a WorkoutViewModel subskrybuje [heartRate], żeby tętno płynęło do treningu
 * niezależnie od cyklu życia ekranu Connect.
 */
@Singleton
class HeartRateManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    companion object {
        private val HR_SERVICE_UUID = UUID.fromString("0000180d-0000-1000-8000-00805f9b34fb")
        private val HR_MEASUREMENT_UUID = UUID.fromString("00002a37-0000-1000-8000-00805f9b34fb")
        private val CLIENT_CHARACTERISTIC_CONFIG_UUID =
            UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
    }

    private var hrGatt: BluetoothGatt? = null

    private val _heartRate = MutableStateFlow<Int?>(null)
    val heartRate: StateFlow<Int?> = _heartRate

    private val _connectionState = MutableStateFlow(BleConnectionState.DISCONNECTED)
    val connectionState: StateFlow<BleConnectionState> = _connectionState

    @SuppressLint("MissingPermission")
    fun connect(device: BluetoothDevice) {
        disconnect()
        _connectionState.value = BleConnectionState.CONNECTING

        hrGatt = device.connectGatt(context, false, object : BluetoothGattCallback() {
            override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
                if (newState == BluetoothGatt.STATE_CONNECTED) {
                    _connectionState.value = BleConnectionState.CONNECTED
                    gatt.discoverServices()
                } else if (newState == BluetoothGatt.STATE_DISCONNECTED) {
                    _connectionState.value = BleConnectionState.DISCONNECTED
                    _heartRate.value = null
                }
            }

            override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
                if (status == BluetoothGatt.GATT_SUCCESS) {
                    val service = gatt.getService(HR_SERVICE_UUID)
                    val characteristic = service?.getCharacteristic(HR_MEASUREMENT_UUID)
                    characteristic?.let {
                        gatt.setCharacteristicNotification(it, true)
                        val descriptor = it.getDescriptor(CLIENT_CHARACTERISTIC_CONFIG_UUID)
                        descriptor?.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                        gatt.writeDescriptor(descriptor)
                    }
                }
            }

            override fun onCharacteristicChanged(
                gatt: BluetoothGatt,
                characteristic: BluetoothGattCharacteristic,
                value: ByteArray
            ) {
                if (characteristic.uuid == HR_MEASUREMENT_UUID) {
                    parseHeartRate(value)?.let { _heartRate.value = it }
                }
            }
        })
    }

    // Standardowy format Heart Rate Measurement: bajt 0 = flagi;
    // bit 0 flag określa, czy wartość HR jest uint8 (bajt 1) czy uint16 (bajty 1-2 LE).
    private fun parseHeartRate(value: ByteArray): Int? {
        if (value.size < 2) return null
        val flags = value[0].toInt() and 0xFF
        return if (flags and 0x01 != 0) {
            if (value.size < 3) null
            else (value[1].toInt() and 0xFF) or ((value[2].toInt() and 0xFF) shl 8)
        } else {
            value[1].toInt() and 0xFF
        }
    }

    @SuppressLint("MissingPermission")
    fun disconnect() {
        hrGatt?.disconnect()
        hrGatt?.close()
        hrGatt = null
        _heartRate.value = null
        _connectionState.value = BleConnectionState.DISCONNECTED
    }
}

File: ios/VBT/VBT/Core/BLE/HeartRateManager.swift
import CoreBluetooth
import Foundation

/// Port 1:1 z Android `data/ble/HeartRateManager.kt` — niezależny od `VbtBleManager`
/// manager standardowego BLE Heart Rate Service (opaska/pas HR, opcjonalny).
@Observable
final class HeartRateManager: NSObject {
    private static let hrServiceUUID = CBUUID(string: "180D")
    private static let hrMeasurementUUID = CBUUID(string: "2A37")

    private(set) var heartRate: Int?
    private(set) var connectionState: BleConnectionState = .disconnected
    private(set) var discoveredPeripherals: [CBPeripheral] = []

    private var central: CBCentralManager!
    private var peripheral: CBPeripheral?

    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: .main)
    }

    func startScan() {
        guard central.state == .poweredOn else { return }
        connectionState = .scanning
        discoveredPeripherals = []
        central.scanForPeripherals(withServices: [Self.hrServiceUUID])
    }

    func stopScan() {
        central.stopScan()
    }

    func connect(to device: CBPeripheral) {
        stopScan()
        connectionState = .connecting
        peripheral = device
        device.delegate = self
        central.connect(device)
    }

    func disconnect() {
        guard let peripheral else {
            connectionState = .disconnected
            return
        }
        central.cancelPeripheralConnection(peripheral)
        heartRate = nil
        connectionState = .disconnected
    }

    /// Standardowy format Heart Rate Measurement: bajt 0 = flagi; bit 0 flag określa,
    /// czy wartość HR jest uint8 (bajt 1) czy uint16 (bajty 1-2 LE).
    private func parseHeartRate(_ value: Data) -> Int? {
        guard value.count >= 2 else { return nil }
        let flags = value[value.startIndex]
        if flags & 0x01 != 0 {
            guard value.count >= 3 else { return nil }
            return Int(value[value.startIndex + 1]) | (Int(value[value.startIndex + 2]) << 8)
        } else {
            return Int(value[value.startIndex + 1])
        }
    }
}

extension HeartRateManager: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {}

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi RSSI: NSNumber) {
        guard !discoveredPeripherals.contains(where: { $0.identifier == peripheral.identifier }) else { return }
        discoveredPeripherals.append(peripheral)
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        connectionState = .connected
        peripheral.discoverServices([Self.hrServiceUUID])
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        connectionState = .disconnected
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        heartRate = nil
        connectionState = .disconnected
    }
}

extension HeartRateManager: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let service = peripheral.services?.first(where: { $0.uuid == Self.hrServiceUUID }) else { return }
        peripheral.discoverCharacteristics([Self.hrMeasurementUUID], for: service)
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard let characteristic = service.characteristics?.first(where: { $0.uuid == Self.hrMeasurementUUID }) else { return }
        peripheral.setNotifyValue(true, for: characteristic)
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard characteristic.uuid == Self.hrMeasurementUUID, let data = characteristic.value else { return }
        if let hr = parseHeartRate(data) {
            heartRate = hr
        }
    }
}
